### PR TITLE
resize opset11

### PR DESCRIPTION
**Pull request by askhade**
**Description**: Implemented all the updates to resize operator per opset 11 spec

**Motivation and Context**
- Adding this change as part of the opset 11 updates

### CODE REVIEW

**Comment by hariharans29** [comment]
I think the relevant tests need to be removed from the exclusion lists...

**Comment by fdwr** [comment]
We currently have some GPU vs GPU ONNX conformance test cases disabled because ORT defaults to coordinate_transformation_mode `asymmetric`, not `half_pixel` (which is the ONNX 1.6 default). After this change, we should be able to re-enable those cases?

    {
      "verification": {"executor_filters": [".*Cpu"]}, // ORT version 0.4 uses coordinate_transformation_mode asymmetric, not half_pixel (the ONNX 1.6 default).
      "graph_name": "Resize 2D wrapped in 4D linear 2x half_pixel",
      "op_type": "Resize",
      "mode": "linear",
      "X": [[[[1, 2],
              [3, 4]]]],
      "scales": {"value": [1, 1, 2, 2], "use_variable_length_values": true},
      "Y": [[[[1,   1.25, 1.75, 2  ],
              [1.5, 1.75, 2.25, 2.5],
              [2.5, 2.75, 3.25, 3.5],
              [3,   3.25, 3.75, 4  ]]]],
      "T": "float32"
    },

**Comment by askhade** [comment]
> We currently have some GPU vs GPU ONNX conformance test cases disabled because ORT defaults to coordinate_transformation_mode `asymmetric`, not `half_pixel` (which is the ONNX 1.6 default). After this change, we should be able to re-enable those cases?
> 
> ```
> {
>   "verification": {"executor_filters": [".*Cpu"]}, // ORT version 0.4 uses coordinate_transformation_mode asymmetric, not half_pixel (the ONNX 1.6 default).
>   "graph_name": "Resize 2D wrapped in 4D linear 2x half_pixel",
>   "op_type": "Resize",
>   "mode": "linear",
>   "X": [[[[1, 2],
>           [3, 4]]]],
>   "scales": {"value": [1, 1, 2, 2], "use_variable_length_values": true},
>   "Y": [[[[1,   1.25, 1.75, 2  ],
>           [1.5, 1.75, 2.25, 2.5],
>           [2.5, 2.75, 3.25, 3.5],
>           [3,   3.25, 3.75, 4  ]]]],
>   "T": "float32"
> },
> ```

Yes! After this change Resize conforms to opset 11 spec which makes half_pixel as default

**Comment by askhade** [comment]
> I think the relevant tests need to be removed from the exclusion lists...

Yes I remember!

File: onnxruntime/core/providers/cpu/tensor/upsample.cc [hunk 1]
```diff
@@ -334,20 +373,188 @@ void upsampleBilinear(
...
+  auto grid_start_pos = (y)*input_width + (x - 1);
+  if (cache.find(grid_start_pos) != cache.end()) {
+    return cache[grid_start_pos];
+  }
+
+  // get the neighbors in 1D and find interpolation for this dimension
+  // for 1D cubic interpolation 4 samples are used. 2 on the left and 2 on the right of x
+  std::array<T, CubicModeGridLength> data_array;
+  float result = 0;
+  for (int i = 0, j = -1; i < (int)CubicModeGridLength; i++, j++) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/upsample.cc]: Use c++ style casts.

File: onnxruntime/core/providers/cpu/tensor/upsample.cc [hunk 2]
```diff
@@ -334,20 +373,188 @@ void upsampleBilinear(
...
+  coeffs[3] = static_cast<float>(((cubic_coeff_a * (2 - abs_s) - 5 * cubic_coeff_a) * (2 - abs_s) + 8 * cubic_coeff_a) * (2 - abs_s) - 4 * cubic_coeff_a);
+  return coeffs;
+}
+
+// Get the tensor data at the requested coordinate
+template <typename T>
+T GetDataForCoordinate(const T* Xdata,
+                       int64_t x, int64_t y,
+                       int64_t input_height, int64_t input_width) {
+  x = std::max((int64_t)0, std::min(x, input_width - 1));
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/upsample.cc]: Use c++ style casts. Below line as well.

File: onnxruntime/core/providers/cpu/tensor/upsample.h [hunk 3]
```diff
@@ -89,6 +228,26 @@ class UpsampleBase {
...
   }
+
+  void ParseScalesDataFromSizes(const Tensor* sizes, const std::vector<int64_t>& dims, std::vector<float>& scales) const {
+    const auto* sizes_data = sizes->template Data<int64_t>();
+    int64_t scales_size = sizes->Shape().Size();
+    if (scales.empty()) {
+      scales.resize(scales_size);
+    }
+    for (size_t i = 0; i < dims.size(); i++) {
+      scales[i] = (float)(sizes_data[i]) / (float)(dims[i]);
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/upsample.h]: Use c++ style casts.

File: onnxruntime/core/providers/cpu/tensor/upsample.h [hunk 4]
```diff
@@ -31,33 +41,162 @@ class UpsampleBase {
...
     if (strcmp(mode.c_str(), UpsampleModeNN) == 0) {
       return UpsampleMode::NN;
     }
     if (strcmp(mode.c_str(), UpsampleModeLinear) == 0) {
       return UpsampleMode::LINEAR;
     }
-      ORT_THROW("mode attribute is " + mode + ". It can only be " +
-                UpsampleModeNN + "(default) or " + UpsampleModeLinear + ".");
+
+    if (strcmp(mode.c_str(), UpsampleModeCubic) == 0) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/upsample.h]: Do we need strcasecmp?

File: onnxruntime/core/providers/cpu/tensor/upsample.cc [hunk 5]
```diff
@@ -250,24 +260,29 @@ Status upsampleLiner(const T* input,
...
-    int64_t batch_size,
-    int64_t num_channels,
-    int64_t input_height,
-    int64_t input_width,
-    float height_scale,
-    float width_scale,
-    const T* Xdata,
-    T* Ydata,
-    AllocatorPtr& alloc) {
+void upsampleBilinear(int64_t batch_size,
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/upsample.cc]: Naming convention. Looks like this was existing code.

File: onnxruntime/core/providers/cpu/tensor/upsample.h [hunk 6]
```diff
@@ -89,6 +228,26 @@ class UpsampleBase {
...
     ScalesValidation(scales, mode_);
   }
+
+  void ParseScalesDataFromSizes(const Tensor* sizes, const std::vector<int64_t>& dims, std::vector<float>& scales) const {
+    const auto* sizes_data = sizes->template Data<int64_t>();
+    int64_t scales_size = sizes->Shape().Size();
+    if (scales.empty()) {
+      scales.resize(scales_size);
+    }
+    for (size_t i = 0; i < dims.size(); i++) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/upsample.h]:
```suggestion
    for (size_t i = 0, end = dims.size(); i < end; ++i) {
```

File: onnxruntime/core/providers/cpu/tensor/upsample.cc [hunk 7]
```diff
@@ -334,20 +373,188 @@ void upsampleBilinear(
...
+
+  // get the neighbors in 1D and find interpolation for this dimension
+  // for 1D cubic interpolation 4 samples are used. 2 on the left and 2 on the right of x
+  std::array<T, CubicModeGridLength> data_array;
+  float result = 0;
+  for (int i = 0, j = -1; i < (int)CubicModeGridLength; i++, j++) {
+    data_array[i] = GetDataForCoordinate(Xdata, x + j, y, input_height, input_width);
+  }
+
+  for (size_t i = 0; i < data_array.size(); i++) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/upsample.cc]:
```suggestion
  for (size_t i = 0. end = data_array.size(); i < end; ++i) {
```

File: onnxruntime/core/providers/cpu/tensor/upsample.cc [hunk 8]
```diff
@@ -334,20 +373,188 @@ void upsampleBilinear(
...
+          std::array<float, CubicModeGridLength> x_interpolation_result;
+          for (int64_t y_val = y_int - 1, i = 0; y_val <= y_int + 2; y_val++, i++) {
+            x_interpolation_result[i] = CubicInterpolation1D(Xdata, x_int, y_val,
+                                                             input_height, input_width, coeff_x,
+                                                             interpolation_result_cache);
+          }
+
+          // From the result of cubic interpolation in x dim, compute cubic interpolation in y dimension
+          float result = 0;
+          for (size_t i = 0; i < x_interpolation_result.size(); i++) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/upsample.cc]:
```suggestion
          for (size_t i = 0, end = x_interpolation_result.size(); i < end; ++i) {
```

File: onnxruntime/core/providers/cpu/tensor/upsample.h (hunk 4 above)
Comment (askhade) [reply onnxruntime/core/providers/cpu/tensor/upsample.h]: Nope... it was part of old code... changed it now

File: onnxruntime/core/providers/cpu/cpu_execution_provider.cc [hunk 9]
```diff
@@ -409,6 +409,7 @@ class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, Ga
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, Range);
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, Unique);
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, TopK);
+class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, float, Resize);
```
Comment (yufenglee) [inline onnxruntime/core/providers/cpu/cpu_execution_provider.cc]: Why only float? Old one also support int8 and int32

File: onnxruntime/core/providers/cpu/tensor/upsample.cc [hunk 10]
```diff
@@ -334,20 +390,180 @@ void upsampleBilinear(
...
+    const std::vector<float>& roi,
+    const T* Xdata,
+    T* Ydata,
+    GetOriginalCoordinateFunc get_original_coordinate) {
+  auto output_width = static_cast<int64_t>(input_width * width_scale);
+  auto output_height = static_cast<int64_t>(input_height * height_scale);
+
+  std::vector<float> y_original;
+  std::vector<float> x_original;
+  std::unordered_map<float, std::array<float, CubicModeGridLength>> cubic_coeffs;
```
Comment (yufenglee) [inline onnxruntime/core/providers/cpu/tensor/upsample.cc]: >cubic_coeffs [](start = 68, length = 12)

emm, i think 2 vector is better here. One for each dimension and use the output_x/y as the index.

**Pull request merged**